pkg/envoy/ads/stream.go

```diff
@@ -185,41 +223,57 @@ func (s *Server) StreamAggregatedResources(server xds_discovery.AggregatedDiscov
 				xdsUpdatePaths = mapset.NewSetWith(typeURL)
 			}
 
-			err = s.sendResponse(xdsUpdatePaths, proxy, &server, &discoveryRequest, s.cfg)
-			if err != nil {
-				log.Error().Err(err).Msgf("Failed to create and send %s update to Envoy with xDS Certificate SerialNumber=%s on Pod with UID=%s",
-					envoy.XDSShortURINames[typeURL], proxy.GetCertificateSerialNumber(), proxy.GetPodUID())
-				continue
-			}
+			jobDone := make(chan struct{}, 1)
+			s.workers.AddJob(&sendJob{
+				typeurls:       xdsUpdatePaths,
+				proxy:          proxy,
+				cfg:            s.cfg,
+				adsStream:      &server,
+				request:        &discoveryRequest,
+				xdsServer:      s,
+				jobDoneChannel: jobDone,
+			})
+			<-jobDone
 
 		case <-broadcastUpdate:
 			log.Info().Msgf("Broadcast wake for Proxy SerialNumber=%s UID=%s", proxy.GetCertificateSerialNumber(), proxy.GetPodUID())
-			err := s.sendResponse(mapset.NewSetWith(
-				envoy.TypeCDS,
-				envoy.TypeEDS,
-				envoy.TypeLDS,
-				envoy.TypeRDS,
-				envoy.TypeSDS),
-				proxy, &server, nil, s.cfg)
-			if err != nil {
-				log.Error().Err(err).Msgf("Failed to create and send ADS update to Envoy with xDS Certificate SerialNumber=%s on Pod with UID=%s",
-					proxy.GetCertificateSerialNumber(), proxy.GetPodUID())
-				continue
-			}
+			jobDone := make(chan struct{}, 1)
+			s.workers.AddJob(&sendJob{
+				typeurls: mapset.NewSetWith(
+					envoy.TypeCDS,
+					envoy.TypeEDS,
+					envoy.TypeLDS,
+					envoy.TypeRDS,
+					envoy.TypeSDS),
+				proxy:          proxy,
+				cfg:            s.cfg,
+				adsStream:      &server,
+				request:        nil,
+				xdsServer:      s,
+				jobDoneChannel: jobDone,
+			})
+			<-jobDone
```

Comment:
If you initialize the `sendJob` struct outside of the evaluation of `AddJob()` you can create the `jobDone` at that time:
```go
job := sendJob{
    jobDoneChannel: make(chan struct{}, 1)
}
```
Then this line here would read `<-job.jobDoneChannel`
And if you rename `jobDoneChannel` to `done` or `isDone`
this would read `<-job.isDone`

Reply:
this is definitely better. will do.